src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java

```diff
@@ -0,0 +1,263 @@
+     *
+     * @return true if this tool requires a reference, otherwise false
+     */
+    public boolean requiresReference() {
+        return false;
+    }
+
+    /**
+     * Does this tool require reads? Tools that do should override to return true.
+     *
+     * @return true if this tool requires reads, otherwise false
+     */
+    public boolean requiresReads() {
+        return false;
+    }
+
+    /**
+     * Does this tool require intervals? Tools that do should override to return true.
+     *
+     * @return true if this tool requires intervals, otherwise false
+     */
+    public boolean requiresIntervals() {
+        return false;
+    }
+
+    /**
+     * Is a source of reference data available?
+     *
+     * @return true if a reference is available, otherwise false
+     */
+    public final boolean hasReference() {
+        return referenceSource != null;
+    }
+
+    /**
+     * Are sources of reads available?
+     *
+     * @return true if reads are available, otherwise false
+     */
+    public final boolean hasReads() {
+        return readsSource != null;
+    }
+
+    /**
+     * Are sources of intervals available?
+     *
+     * @return true if intervals are available, otherwise false
+     */
+    public final boolean hasIntervals() {
+        return intervals != null;
+    }
+
+    /**
+     * Window function that controls how much reference context to return for each read when
+     * using the reference source returned by {@link #getReference}. Tools should override
+     * as appropriate. The default function is the identity function (ie., return exactly
+     * the reference bases that span each read).
+     *
+     * @return reference window function used to initialize the reference source
+     */
+    public SerializableFunction<GATKRead, SimpleInterval> getReferenceWindowFunction() {
+        return RefWindowFunctions.IDENTITY_FUNCTION;
+    }
+
+    /**
+     * Returns the "best available" sequence dictionary. This will be the reference sequence dictionary if
+     * there is a reference, otherwise it will be the sequence dictionary constructed from the reads if
+     * there are reads, otherwise it will be null.
+     *
+     * TODO: check interval file(s) as well for a sequence dictionary
+     *
+     * @return best available sequence dictionary given our inputs
+     */
+    public SAMSequenceDictionary getBestAvailableSequenceDictionary() {
+        return hasReference() ? referenceDictionary : (hasReads() ? readsHeader.getSequenceDictionary() : null);
+    }
+
+    /**
+     * @return sequence dictionary for the reference, or null if there is no reference available
+     */
+    public SAMSequenceDictionary getReferenceSequenceDictionary() {
+        return referenceDictionary;
+    }
+
+    /**
+     * @return header for the reads, or null if there are no reads available
+     */
+    public SAMFileHeader getHeaderForReads() {
```

Comment:
I think it depends on how "null-tolerant" our final interfaces end up being. Right now some things work transparently with null (eg., `getAuthenticatedGCSOptions()`) and some don't, so I think we should wait a bit before deciding whether an exception-based approach is the way to go when missing data is requested.